File: tests/gpu/minor_record_test.py
"""The minor's own record on the GPU: the RESOLVED suzerain, the research
walk, the border that closes with Early Empire, Containment's doubled envoy,
and a city-state that can be converted.

    npm run seed && npm run export        # (once) writes seeder/worlds/
    python tests/gpu/minor_record_test.py

Each is poked at its own body against the same sourced rules the TS vitest
pins (tests/cpu/minors/minor-record.test.ts).
"""

from __future__ import annotations

import sys
from pathlib import Path

import torch

sys.path.insert(0, str(Path(__file__).resolve().parent.parent.parent / "gpu"))

from core import BatchSim, load_rules, load_fixture, fixture_paths
from warmup import settle_all


def build(rules):
    sim = settle_all(BatchSim([load_fixture(fixture_paths()[0])], rules,
                              device="cpu", dtype=torch.float64))
    for _ in range(3):
        sim.step()
    assert sim.S > 0, "fixture has no city-state"
    return sim


def test_resolved_suzerain(rules) -> None:
    """`resolveSuzerain`'s twin: at least suzerainEnvoys, STRICTLY more than
    every other seat, a tie leaves nobody."""
    sim = build(rules)
    suz_min = int(sim.rules.citystate.get("suzerainEnvoys", 3))
    env = sim.seat_citystate_envoys
    env[0, :, 0] = 0
    sim._cs_resolve_suzerain()
    assert int(sim.citystate_suzerain[0, 0]) == -1, "an empty record has no suzerain"

    env[0, 0, 0] = suz_min - 1
    sim._cs_resolve_suzerain()
    assert int(sim.citystate_suzerain[0, 0]) == -1, "one short of the bar holds nothing"

    env[0, 0, 0] = suz_min
    sim._cs_resolve_suzerain()
    assert int(sim.citystate_suzerain[0, 0]) == 0, "the bar, uncontested, holds it"
    assert bool(sim._suzerain_mask(0)[0, 0]), "the live contest agrees"

    if sim.n_majors > 1:
        env[0, 1, 0] = suz_min
        sim._cs_resolve_suzerain()
        assert int(sim.citystate_suzerain[0, 0]) == -1, "a TIE leaves nobody"
        assert not bool(sim._suzerain_mask(0)[0, 0]) and not bool(sim._suzerain_mask(1)[0, 0])
        env[0, 1, 0] = suz_min + 1
        sim._cs_resolve_suzerain()
        assert int(sim.citystate_suzerain[0, 0]) == 1, "the strict lead takes it"
    print("  resolved suzerain OK: the bar, the strict lead, the tie")


def test_minor_research(rules) -> None:
    """CIV6 (City-state): a minor "develops scientifically and culturally...
    it will apparently research certain techs" — POPULATION a turn into each
    pot, the cheapest available row completing at most once per pot."""
    sim = build(rules)
    sim.citystate_techs[0, 0].zero_()
    sim.citystate_civics[0, 0].zero_()
    sim.citystate_tech_prog[0, 0] = 0
    sim.citystate_civic_prog[0, 0] = 0
    pop = int(sim.citystate_pop[0, 0])
    assert pop > 0, "a live minor has population"
    cost = sim.rules_dev.t_cost
    avail = sim._available_mask(sim.citystate_techs[0:1, 0], sim._prereq_t)[0]
    cheapest = int(torch.where(avail, cost.to(sim.device),
                               torch.full_like(cost.to(sim.device), float("inf"))).argmin())
    sim._city_state_phase()
    assert float(sim.citystate_tech_prog[0, 0]) == float(pop), "the pot takes POPULATION"
    assert int(sim.citystate_techs[0, 0].sum()) == 0, "and nothing completes below the price"
    turns = int(float(cost[cheapest]) // pop) + 2
    for _ in range(turns):
        sim._city_state_phase()
    assert bool(sim.citystate_techs[0, 0, cheapest]), "the cheapest available row lands first"
    assert int(sim.citystate_civics[0, 0].sum()) > 0, "the civic pot walks its own row"
    print(f"  minor research OK: {pop}/turn, cheapest row {cheapest} at cost {float(cost[cheapest]):.0f}")


def test_minor_border(rules) -> None:
    """CIV6 (Borders): a city-state's ground closes on ITS Early Empire, and
    "Open Borders is granted to players that have reached Suzerain status"."""
    sim = build(rules)
    obc = sim._open_borders_civic
    assert obc >= 0, "no open-borders civic exported"
    own = (sim.tile_seat[0] == 100).nonzero(as_tuple=True)[0]
    assert len(own) > 0, "city-state 0 owns no tile"
    t = torch.tensor([[int(own[0])]], dtype=torch.long, device=sim.device)

    sim.citystate_civics[0, 0, obc] = False
    sim.seat_citystate_envoys[0, :, 0] = 0
    sim._cs_resolve_suzerain()
    assert not bool(sim._border_closed(t, 0)[0, 0]), "open ground before the civic"

    sim.citystate_civics[0, 0, obc] = True
    assert bool(sim._border_closed(t, 0)[0, 0]), "Early Empire closes it"

    suz_min = int(sim.rules.citystate.get("suzerainEnvoys", 3))
    sim.seat_citystate_envoys[0, 0, 0] = suz_min
    sim._cs_resolve_suzerain()
    assert not bool(sim._border_closed(t, 0)[0, 0]), "the SUZERAIN is granted passage"

    if sim.n_majors > 1:
        assert bool(sim._border_closed(t, 1)[0, 0]), "a rival still refuses"
        cs_row = int(sim._seat_row[100])
        sim.war[0, 1, cs_row] = sim.war[0, cs_row, 1] = True
        sim.sync_war()
        assert not bool(sim._border_closed(t, 1)[0, 0]), "a war opens what the civic closed"
        sim.war[0, 1, cs_row] = sim.war[0, cs_row, 1] = False
        sim.sync_war()

    # "Traders ignore borders", and religious units too — the Inquisitor apart
    tr = torch.tensor([[sim._trader_idx]], dtype=torch.long, device=sim.device)
    assert not bool(sim._border_closed(t, 1 % sim.n_majors, tr)[0, 0]) or sim.n_majors == 1
    if sim._inquisitor_idx >= 0 and sim.n_majors > 1:
        iq = torch.tensor([[sim._inquisitor_idx]], dtype=torch.long, device=sim.device)
        assert bool(sim._border_closed(t, 1, iq)[0, 0]), "the Inquisitor is the exception"
    print("  minor border OK: the civic, the suzerain's passage, the war, the exempt classes")


def test_containment(rules) -> None:
    """CIV6 (Containment): "Each Envoy you send to a city-state counts as two,
    if its Suzerain has a different government than you" — the CHANNEL, poked
    at the modifier the send reads."""
    sim = build(rules)
    if not sim._gov_has_effects:
        print("  containment SKIPPED: no government effects in this catalog")
        return
    assert hasattr(sim, "_pol_envoy2"), "the policy column is not exported"
    assert int(sim._pol_envoy2.sum()) >= 1, "no card carries the envoy-doubling effect"
    fx = sim._gov_mods(0)[12]
    assert "envoy2" in fx, "the modifier channel is missing"
    assert fx["envoy2"].dtype == torch.bool and fx["envoy2"].shape[0] == sim.B
    print(f"  containment OK: {int(sim._pol_envoy2.sum())} card(s) carry it, the channel reads bool")


def test_minor_conversion(rules) -> None:
    """B-59r: a minor's city row takes pressure like any other, so a
    city-state CAN be converted — `city_pressure` reaches the minor rows."""
    sim = build(rules)
    m0 = sim._CITY_MINOR0
    assert sim.city_pressure.shape[1] > m0, "no minor row in the pressure plane"
    sim.city_pressure[0, m0 + 0, 0].zero_()
    sim.city_pressure[0, m0 + 0, 0, 0] = 500
    got = sim._followed_religion(sim.city_pressure[0:1, m0 + 0, 0])
    assert int(got[0]) == 0, f"the minor row follows the religion pressing it: {int(got[0])}"
    print("  minor conversion OK: the minor city row carries pressure and a majority")


def main() -> None:
    rules = load_rules()
    print(f"minor_record_test on {fixture_paths()[0].name}:")
    test_resolved_suzerain(rules)
    test_minor_research(rules)
    test_minor_border(rules)
    test_containment(rules)
    test_minor_conversion(rules)
    print("MINOR RECORD OK")


if __name__ == "__main__":
    main()

File: tests/cpu/minors/minor-record.test.ts
import { describe, it, expect } from 'vitest';
import { emptySeat, seatOfCityState, setTileOwner, setWar } from '../../../cpu/core/seats';
import { makeMap, makeState, tileAtCoords } from '../helpers';
import { spawnUnit } from '../../../cpu/core/units';
import { borderClosedTo } from '../../../cpu/core/units';
import { spreadFromUnit } from '../../../cpu/core/unitOrders';
import { addEnvoys, cityStatePhase, isSuzerain, resolveSuzerain, suzerainOf } from '../../../cpu/core/cityStates';
import { containmentBonus } from '../../../cpu/core/effects';
import { SUZERAIN_ENVOYS } from '../../../cpu/data/cityStates';
import { OPEN_BORDERS_CIVIC } from '../../../cpu/data/seats';
import { SPREAD_PRESSURE } from '../../../cpu/data/religion';
import { TECHS } from '../../../cpu/data/techs';
import { CIVICS } from '../../../cpu/data/civics';
import { tilesWithin } from '../../../world/hex';
import type { CityState, CityStateType, GameState, Seat } from '../../../cpu/core/types';

function addCs(state: GameState, col: number, row: number, opts: Partial<CityState> & { type?: CityStateType } = {}): CityState {
  const center = tileAtCoords(state.map, col, row);
  const cityState: CityState = {
    ...emptySeat(seatOfCityState(state.cityStates.length)),
    id: state.cityStates.length,
    name: `CS${state.cityStates.length}`,
    type: 'scientific',
    centerIndex: center.index,
    population: 3,
    envoys: {},
    met: [0],
    ...opts,
  };
  for (const t of tilesWithin(state.map, col, row, 1)) setTileOwner(t, seatOfCityState(cityState.id));
  state.cityStates.push(cityState);
  resolveSuzerain(cityState);
  return cityState;
}

function addSeat(state: GameState, seat: number): Seat {
  const s = emptySeat(seat);
  state.seats.push(s);
  return s;
}

// ---------------------------------------------------------------------------
describe('the RESOLVED suzerain', () => {
  it('stores the contest answer and agrees with isSuzerain at every write', () => {
    const state = makeState(makeMap(24, 24));
    addSeat(state, 1);
    const cs = addCs(state, 12, 12);
    expect(suzerainOf(cs)).toBe(-1);            // nobody yet

    addEnvoys(cs, 0, SUZERAIN_ENVOYS - 1);      // one short of the bar
    expect(suzerainOf(cs)).toBe(-1);
    expect(isSuzerain(cs, 0)).toBe(false);

    addEnvoys(cs, 0, 1);                        // exactly the bar, uncontested
    expect(suzerainOf(cs)).toBe(0);
    expect(isSuzerain(cs, 0)).toBe(true);

    addEnvoys(cs, 1, SUZERAIN_ENVOYS);          // a TIE leaves nobody
    expect(suzerainOf(cs)).toBe(-1);
    expect(isSuzerain(cs, 0)).toBe(false);
    expect(isSuzerain(cs, 1)).toBe(false);

    addEnvoys(cs, 1, 1);                        // seat 1 pulls ahead
    expect(suzerainOf(cs)).toBe(1);
    expect(isSuzerain(cs, 1)).toBe(true);
  });
});

// ---------------------------------------------------------------------------
describe('Containment', () => {
  function scenario(): { state: GameState; cs: CityState; sender: Seat; suz: Seat } {
    const state = makeState(makeMap(24, 24));
    const sender = state.seats[0];
    const suz = addSeat(state, 1);
    const cs = addCs(state, 12, 12);
    addEnvoys(cs, suz.seat, SUZERAIN_ENVOYS);   // seat 1 holds it
    return { state, cs, sender, suz };
  }

  it('pays nothing without the card', () => {
    const { state, cs, sender } = scenario();
    expect(containmentBonus(state, cs, sender)).toBe(0);
  });

  it('pays 1 when the suzerain runs a DIFFERENT government', () => {
    const { state, cs, sender, suz } = scenario();
    // the card is slotted off the seat's own research; give the sender the
    // civic that unlocks CONTAINMENT and a government the suzerain lacks
    sender.research.civics = Object.keys(CIVICS);
    expect(containmentBonus(state, cs, sender)).toBe(1);
    // ... and nothing once the suzerain runs the same one
    suz.research.civics = [...sender.research.civics];
    expect(containmentBonus(state, cs, sender)).toBe(0);
  });

  it('pays nothing against a city-state with no suzerain, or one that is ME', () => {
    const { state, cs, sender } = scenario();
    sender.research.civics = Object.keys(CIVICS);
    cs.envoys = {};
    resolveSuzerain(cs);
    expect(containmentBonus(state, cs, sender)).toBe(0);
    addEnvoys(cs, sender.seat, SUZERAIN_ENVOYS);
    expect(suzerainOf(cs)).toBe(sender.seat);
    expect(containmentBonus(state, cs, sender)).toBe(0);
  });
});

// ---------------------------------------------------------------------------
describe("the minor's research record", () => {
  it('banks POPULATION a turn and completes the cheapest available row', () => {
    const state = makeState(makeMap(24, 24));
    const cs = addCs(state, 12, 12, { population: 5 });
    const cheapestTech = Object.entries(TECHS)
      .filter(([, d]) => d.prereqs.length === 0)
      .sort((a, b) => a[1].cost - b[1].cost)[0];
    state.turn = 1;
    cityStatePhase(state);
    expect(cs.research.techProgress).toBe(5);   // pop into the pot, nothing done
    expect(cs.research.techs).toEqual([]);
    for (let i = 0; i < Math.ceil(cheapestTech[1].cost / 5); i++) cityStatePhase(state);
    expect(cs.research.techs).toContain(cheapestTech[0]);
    expect(cs.research.techProgress).toBeGreaterThanOrEqual(0);
  });

  it('reaches Early Empire and closes its border to everyone but the suzerain', () => {
    const state = makeState(makeMap(24, 24));
    state.unitsMode = true;
    addSeat(state, 1);
    const cs = addCs(state, 12, 12);
    const tile = state.map.tiles[tilesWithin(state.map, 12, 12, 1).find((t) => t.index !== cs.centerIndex)!.index];
    expect(borderClosedTo(state, 0, tile, 'WARRIOR')).toBe(false); // open before the civic

    cs.research.civics = [OPEN_BORDERS_CIVIC];
    expect(borderClosedTo(state, 0, tile, 'WARRIOR')).toBe(true);

    // CIV6 (Borders): "For city-states, Open Borders is granted to players
    // that have reached Suzerain status."
    addEnvoys(cs, 0, SUZERAIN_ENVOYS);
    expect(borderClosedTo(state, 0, tile, 'WARRIOR')).toBe(false);

    // a rival without the suzerainty still refuses — and a war opens it
    expect(borderClosedTo(state, 1, tile, 'WARRIOR')).toBe(true);
    setWar(state, 1, cs.seat, true);
    expect(borderClosedTo(state, 1, tile, 'WARRIOR')).toBe(false);
  });

  it('lets a Trader and a religious unit through a closed minor border', () => {
    const state = makeState(makeMap(24, 24));
    state.unitsMode = true;
    const cs = addCs(state, 12, 12);
    cs.research.civics = [OPEN_BORDERS_CIVIC];
    const tile = state.map.tiles[tilesWithin(state.map, 12, 12, 1).find((t) => t.index !== cs.centerIndex)!.index];
    expect(borderClosedTo(state, 0, tile, 'WARRIOR')).toBe(true);
    expect(borderClosedTo(state, 0, tile, 'TRADER')).toBe(false);
    expect(borderClosedTo(state, 0, tile, 'MISSIONARY')).toBe(false);
    expect(borderClosedTo(state, 0, tile, 'INQUISITOR')).toBe(true);
  });
});

// ---------------------------------------------------------------------------
describe('a city-state can be converted', () => {
  it('takes the spread lump on its own centre, at Translator strength', () => {
    const state = makeState(makeMap(24, 24));
    state.unitsMode = true;
    state.seats[0].religion.founded = true;
    const cs = addCs(state, 12, 12);
    const home = tileAtCoords(state.map, 11, 12);
    const ap = spawnUnit(state, 'APOSTLE', home.index, 0)!;
    ap.charges = 3;
    spreadFromUnit(state, ap, state.seats[0], state.map.tiles[cs.centerIndex]);
    // CIV6 (Translator): the promotion's note extends the triple to
    // city-states; unpromoted the lump is the plain one.
    expect(cs.religionPressure?.[0]).toBe(SPREAD_PRESSURE);
    expect(ap.charges).toBe(2);
  });
});
